# IRIS DATASET

In [168]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

iris = pd.read_csv("Iris.csv")

iris.head()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa


In [169]:
# Remove Id column if exists
if 'Id' in iris.columns:
    iris.drop('Id', axis=1, inplace=True)

# Features and target
X = iris.drop('Species', axis=1)

le = LabelEncoder()
y = le.fit_transform(iris['Species'])

In [170]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [171]:
X.head()

,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


In [172]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [173]:
print("Dataset Shape:", X.shape)
print("Number of Classes:", len(np.unique(y)))

Dataset Shape: (150, 4)
Number of Classes: 3


### Calculate Metrics

In [174]:
def evaluate_model(model, X_train, X_test, y_train, y_test):

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)

    classification_error = 1 - accuracy

    precision = precision_score(
        y_test,
        y_pred,
        average='weighted'
    )

    recall = recall_score(
        y_test,
        y_pred,
        average='weighted'
    )

    f1 = f1_score(
        y_test,
        y_pred,
        average='weighted'
    )

    return {
        "Classification Error": classification_error,
        "Precision": precision,
        "Recall": recall,
        "F1": f1
    }



## K-NN

In [175]:
k_values = [1, 3, 5, 7, 9, 11, 13, 15]

knn_results = []

for k in k_values:

    knn = KNeighborsClassifier(n_neighbors=k)

    metrics = evaluate_model(
        knn,
        X_train,
        X_test,
        y_train,
        y_test
    )

    knn_results.append([
        k,
        metrics["Classification Error"],
        metrics["Precision"],
        metrics["Recall"],
        metrics["F1"]
    ])

knn_df = pd.DataFrame(
    knn_results,
    columns=[
        "k",
        "Classification Error",
        "Precision",
        "Recall",
        "F1"
    ]
)

print(knn_df)

    k  Classification Error  Precision    Recall        F1
0   1              0.033333   0.969697  0.966667  0.966583
1   3              0.000000   1.000000  1.000000  1.000000
2   5              0.000000   1.000000  1.000000  1.000000
3   7              0.033333   0.969697  0.966667  0.966583
4   9              0.000000   1.000000  1.000000  1.000000
5  11              0.033333   0.969697  0.966667  0.966583
6  13              0.033333   0.969697  0.966667  0.966583
7  15              0.033333   0.969697  0.966667  0.966583


### Conclusion

- The classifier achieved perfect performance at k = 3, 5, and 9, with:
    - Classification Error = 0
    - Precision = 1.0
    - Recall = 1.0
    - F1-score = 1.0


- The values k = 1 and k ≥ 7 produced slightly lower but identical performance metrics.

- Although k = 1 and larger k values behaved similarly on this dataset, they represent different classification strategies. The similarity in results suggests that the dataset is relatively easy to classify and is not highly sensitive to the choice of k.

#### Normalize Features

In [176]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

In [177]:
knn_scaled_results = []

for k in k_values:

    knn = KNeighborsClassifier(n_neighbors=k)

    metrics = evaluate_model(
        knn,
        X_train_scaled,
        X_test_scaled,
        y_train,
        y_test
    )

    knn_scaled_results.append([
        k,
        metrics["Classification Error"],
        metrics["Precision"],
        metrics["Recall"],
        metrics["F1"]
    ])

knn_scaled_df = pd.DataFrame(
    knn_scaled_results,
    columns=[
        "k",
        "Classification Error",
        "Precision",
        "Recall",
        "F1"
    ]
)

print(knn_scaled_df)

    k  Classification Error  Precision    Recall        F1
0   1              0.033333   0.969697  0.966667  0.966583
1   3              0.066667   0.944444  0.933333  0.932660
2   5              0.066667   0.944444  0.933333  0.932660
3   7              0.033333   0.969697  0.966667  0.966583
4   9              0.033333   0.969697  0.966667  0.966583
5  11              0.033333   0.969697  0.966667  0.966583
6  13              0.033333   0.969697  0.966667  0.966583
7  15              0.033333   0.969697  0.966667  0.966583


#### Conclusion

Before Normalization
- Best performance at k = 3, 5, and 9
- Classification Error = 0
- Precision = Recall = F1 = 1.0

After Normalization
- Performance became worse for k = 3 and 5
- Classification Error increased to 6.67%
- F1-score decreased from 1.0 to 0.933


Normalization did not improve the performance of the K-NN classifier on the Iris dataset.

Before normalization, the model achieved perfect classification for k = 3, 5, and 9. After normalization, the classification error increased from 0% to 6.67% for k = 3 and 5, while precision, recall, and F1-score also decreased.

This result suggests that the original feature scales already provided useful information for class separation. Since the Iris dataset is relatively clean and well-structured, standardizing the features was not necessary and slightly reduced the effectiveness of K-NN. Therefore, for this dataset, the non-normalized version performed better.

## Logistic Regression

In [178]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(max_iter=5000)

lr_metrics = evaluate_model(
    lr,
    X_train,
    X_test,
    y_train,
    y_test
)

lr_df = pd.DataFrame([lr_metrics])

print(lr_df)

   Classification Error  Precision    Recall        F1
0              0.033333   0.969697  0.966667  0.966583


#### Run with Normalize Features

In [179]:
lr_scaled = LogisticRegression(max_iter=5000)

lr_scaled_metrics = evaluate_model(
    lr_scaled,
    X_train_scaled,
    X_test_scaled,
    y_train,
    y_test
)

lr_scaled_df = pd.DataFrame([lr_scaled_metrics])

print(lr_scaled_df)


   Classification Error  Precision    Recall        F1
0              0.066667   0.933333  0.933333  0.933333


Normalization did not improve the performance of Logistic Regression on the Iris dataset. 

The original model achieved a classification error of 3.33% and an F1-score of 0.967. 

After normalization, the classification error increased to 6.67% and the F1-score decreased to 0.933. 

Therefore, the non-normalized Logistic Regression model performed slightly better on this dataset.

# SVM

In [180]:
from sklearn.svm import SVC

svm = SVC(kernel='rbf')

svm_metrics = evaluate_model(
    svm,
    X_train,
    X_test,
    y_train,
    y_test
)

svm_df = pd.DataFrame([svm_metrics])

print(svm_df)

   Classification Error  Precision    Recall        F1
0              0.033333   0.969697  0.966667  0.966583


#### Run with Normalize Features

In [181]:
svm_scaled = SVC(kernel='rbf')

svm_scaled_metrics = evaluate_model(
    svm_scaled,
    X_train_scaled,
    X_test_scaled,
    y_train,
    y_test
)

svm_scaled_df = pd.DataFrame([svm_scaled_metrics])

print(svm_scaled_df)

   Classification Error  Precision    Recall        F1
0              0.033333   0.969697  0.966667  0.966583


- Both versions of SVM produced exactly the same results.
- Classification Error remained at 3.33%.
- Precision, Recall, and F1-score remained unchanged.

Normalization had no noticeable effect on SVM performance for the Iris dataset. Both models achieved identical results.

# Breast Cancer Wisconsin DATASET

In [182]:
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()

X = data.data
y = data.target

In [183]:
print("Dataset Shape:", X.shape)
print("Number of Classes:", len(np.unique(y)))

# ============================================
# Train/Test Split
# ============================================

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


Dataset Shape: (569, 30)
Number of Classes: 2


## K-NN

In [184]:
k_values = [1, 3, 5, 7, 9, 11, 13, 15]

knn_results = []

for k in k_values:

    knn = KNeighborsClassifier(n_neighbors=k)

    metrics = evaluate_model(
        knn,
        X_train,
        X_test,
        y_train,
        y_test
    )

    knn_results.append([
        k,
        metrics["Classification Error"],
        metrics["Precision"],
        metrics["Recall"],
        metrics["F1"]
    ])

knn_df = pd.DataFrame(
    knn_results,
    columns=[
        "k",
        "Classification Error",
        "Precision",
        "Recall",
        "F1"
    ]
)

print(knn_df)

    k  Classification Error  Precision    Recall        F1
0   1              0.078947   0.923417  0.921053  0.921574
1   3              0.070175   0.929825  0.929825  0.929825
2   5              0.087719   0.913671  0.912281  0.912683
3   7              0.070175   0.929825  0.929825  0.929825
4   9              0.061404   0.938435  0.938596  0.938438
5  11              0.061404   0.938435  0.938596  0.938438
6  13              0.052632   0.947440  0.947368  0.947087
7  15              0.070175   0.929694  0.929825  0.929449


Unlike the Iris dataset, the Breast Cancer Wisconsin dataset shows a clearer relationship between the value of k and classification performance.

The value of k has a noticeable impact on the K-NN classifier for the Breast Cancer Wisconsin dataset. 

The best result was obtained at k = 13, achieving the lowest classification error (5.26%) and the highest F1-score (0.947). 

Smaller values of k were more sensitive to local variations in the data, while very large values began to reduce performance.

Therefore, k = 13 provides the best balance between robustness and classification accuracy for this dataset.


#### Normalize Features

In [185]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

In [186]:
knn_scaled_results = []

for k in k_values:

    knn = KNeighborsClassifier(n_neighbors=k)

    metrics = evaluate_model(
        knn,
        X_train_scaled,
        X_test_scaled,
        y_train,
        y_test
    )

    knn_scaled_results.append([
        k,
        metrics["Classification Error"],
        metrics["Precision"],
        metrics["Recall"],
        metrics["F1"]
    ])

knn_scaled_df = pd.DataFrame(
    knn_scaled_results,
    columns=[
        "k",
        "Classification Error",
        "Precision",
        "Recall",
        "F1"
    ]
)

print(knn_scaled_df)

    k  Classification Error  Precision    Recall        F1
0   1              0.061404   0.939042  0.938596  0.938743
1   3              0.017544   0.982930  0.982456  0.982362
2   5              0.043860   0.956073  0.956140  0.956027
3   7              0.026316   0.974737  0.973684  0.973465
4   9              0.026316   0.974737  0.973684  0.973465
5  11              0.026316   0.974737  0.973684  0.973465
6  13              0.026316   0.974737  0.973684  0.973465
7  15              0.026316   0.974737  0.973684  0.973465


Normalization significantly improved the performance of the K-NN classifier on the Breast Cancer Wisconsin dataset. 

Before normalization, the best model achieved a classification error of 5.26% and an F1-score of 0.947. 

After normalization, the classification error decreased to 1.75% and the F1-score increased to 0.982. 

This improvement occurs because K-NN relies on distance calculations, and normalization prevents features with larger numerical ranges from dominating the distance metric.

## Logistic Regression

In [187]:
lr = LogisticRegression(max_iter=5000)

lr_metrics = evaluate_model(
    lr,
    X_train,
    X_test,
    y_train,
    y_test
)

lr_df = pd.DataFrame([lr_metrics])

print(lr_df)

   Classification Error  Precision    Recall        F1
0              0.035088   0.965185  0.964912  0.964725


#### Run with Normalize Features

In [188]:
lr_scaled = LogisticRegression(max_iter=5000)

lr_scaled_metrics = evaluate_model(
    lr_scaled,
    X_train_scaled,
    X_test_scaled,
    y_train,
    y_test
)

lr_scaled_df = pd.DataFrame([lr_scaled_metrics])

print(lr_scaled_df)

   Classification Error  Precision    Recall        F1
0              0.017544   0.982456  0.982456  0.982456


Normalization improved the performance of Logistic Regression on the Breast Cancer Wisconsin dataset. 

The classification error decreased from 3.51% to 1.75%, 

while precision, recall, and F1-score increased from approximately 96.5% to 98.2%. 

Logistic Regression relies on an optimization algorithm to learn model parameters, and feature scaling helps the algorithm converge more efficiently and find a better decision boundary. Therefore, normalization is beneficial for this dataset.

## SVM

In [189]:
svm = SVC(kernel='rbf')

svm_metrics = evaluate_model(
    svm,
    X_train,
    X_test,
    y_train,
    y_test
)

svm_df = pd.DataFrame([svm_metrics])

print(svm_df)

   Classification Error  Precision    Recall        F1
0              0.070175   0.930748  0.929825  0.929018


#### Run with Normalize Features

In [191]:
svm_scaled = SVC(kernel='rbf')

svm_scaled_metrics = evaluate_model(
    svm_scaled,
    X_train_scaled,
    X_test_scaled,
    y_train,
    y_test
)

svm_scaled_df = pd.DataFrame([svm_scaled_metrics])

print(svm_scaled_df)


   Classification Error  Precision    Recall        F1
0              0.017544   0.982456  0.982456  0.982456


Normalization had a significant positive impact on the SVM classifier.

- Classification Error decreased from 7.02% to 1.75%.
- Precision increased from 93.07% to 98.25%.
- Recall increased from 92.98% to 98.25%.
- F1-score increased from 92.90% to 98.25%.

The normalized SVM model achieved the same excellent performance as the normalized Logistic Regression model.